packages/steiger/src/shared/globs.ts

```diff
@@ -1,3 +1,67 @@
+import { minimatch } from 'minimatch'
+
+// ! Don't use platform specific path separators in the glob patterns for globby/minimatch
+// as it only works with forward slashes!
+
+interface ApplyGlobsOptions {
+  inclusions?: string[]
+  exclusions?: string[]
+}
+
+export type NodeType = 'file' | 'folder'
+
 export function isNegationPattern(pattern: string) {
   return pattern.startsWith('!')
 }
+
+// This function is a wrapper around minimatch to be able
+// to match the containing folder when the pattern ends with `**`.
+// E.g. if the pattern is `/src/shared/**` it should match `/src/shared` as well.
+// But it's not the case with default minimatch.
+function matchWithContainingFolder(path: string, pattern: string, nodeType: NodeType) {
+  const isAllInsideFolderPattern = pattern.endsWith('**')
+
+  const defaultResult = minimatch(path, pattern)
+
+  if (defaultResult) {
+    return defaultResult
+  }
+
+  // If the default result failed, try to match the containing folder
+  if (nodeType === 'folder' && isAllInsideFolderPattern) {
+    return minimatch(path, pattern.slice(0, -3))
+  }
+
+  return false
+}
```

Comment:
issue: I think there are edge cases that this logic doesn't cover. Ideally, we'd want to work like Git — if there are no files in a folder, the folder doesn't exist. The edge case here would be like this:
- 📂 folder1
  - 📄 index.ts
  
and an ignoring glob of `**/index.ts`. This glob would leave folder1 empty, and this logic will not catch that.

suggestion: let's instead change the logic of the function that recomposes a flat VFS back into a tree to do a pass at the end that would remove all empty folders. That should be more robust and shouldn't require us to add any fields to the diagnostic.

Reply:
Do we need this tweaked behavior only when we remove nodes? Because my goal was to make all globs work in the same way in all cases, so they are predictable. So, when we apply error/warn severities it matches the containing folder too, and that is why I put that logic on such a low level

Reply:
What I implemented was a kind of "local" containing folder matching where I matched it for a single glob. I had an idea of "global" matching and it was similar to what you proposed. The idea is to add another tree traversal at the end of severity processing to check if all nodes inside a folder have some type of severity and if it's true we mark the containing folder with the same severity. But it still has flaws as some parts of FS can be remarked, and that won't allow us to definitely say how to mark the containing folder

```javascript
// Here everything inside /lib is off except for get-query-params, which is 'warn'
const config = [
  {
    files: ['./src/shared/lib/**'],
    rules: {
       'fsd/forbidden-imports': 'off',
    }
  },
  {
    files: ['./src/shared/lib/get-query-params.js'],
    rules: {
       'fsd/forbidden-imports': 'warn',
    }
  }
]
```